File: gateway/auth/ts/src/jcs.ts
/**
 * JSON Canonicalization Scheme (RFC 8785) wrapper.
 *
 * Single source of truth for the canonical-JSON step. All signing
 * inputs and HMAC inputs run through this exactly once.
 *
 * Returns the canonical JSON STRING (not bytes). Use `utf8Bytes` to
 * convert when feeding to crypto primitives.
 */

// canonicalize is shipped as CJS with `module.exports = function`; under
// NodeNext + esModuleInterop the default import resolves to that function.
// The package's own `.d.ts` declares `export default function serialize`.
import canonicalizeNs from "canonicalize";

import { utf8Bytes } from "./encoding.js";

type Canonicalize = (input: unknown) => string | undefined;

// Some TS module-resolution settings see the namespace rather than the
// default; tolerate both.
const canonicalize: Canonicalize =
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  typeof (canonicalizeNs as unknown) === "function"
    ? (canonicalizeNs as unknown as Canonicalize)
    : ((canonicalizeNs as unknown as { default: Canonicalize }).default);

export function jcs(value: unknown): string {
  const out = canonicalize(value);
  if (typeof out !== "string") {
    throw new Error(
      "canonicalize() returned non-string (input may contain undefined or non-JSON types)",
    );
  }
  return out;
}

/**
 * Strip one named field from a layer object, then canonicalize the
 * rest, then return UTF-8 bytes. This is the "signing input" recipe
 * from phase 4.
 */
export function signingBytes(
  layer: Record<string, unknown>,
  sigFieldName: string,
): Uint8Array {
  const stripped: Record<string, unknown> = {};
  for (const key of Object.keys(layer)) {
    if (key !== sigFieldName) {
      stripped[key] = layer[key];
    }
  }
  return utf8Bytes(jcs(stripped));
}

File: gateway/auth/go/jcs.go
package macaroon

import (
	"encoding/json"
	"fmt"

	"github.com/gowebpki/jcs"
)

// JCS canonicalizes a Go value to RFC 8785 canonical JSON.
//
// The value is first marshalled with encoding/json (standard library)
// then transformed by github.com/gowebpki/jcs to RFC 8785 form. The
// result is the canonical UTF-8 JSON byte sequence — feed it directly
// to signers / HMAC primitives, no further encoding.
func JCS(v any) ([]byte, error) {
	raw, err := json.Marshal(v)
	if err != nil {
		return nil, fmt.Errorf("json.Marshal: %w", err)
	}
	out, err := jcs.Transform(raw)
	if err != nil {
		return nil, fmt.Errorf("jcs.Transform: %w", err)
	}
	return out, nil
}

// JCSStripField marshals v to JSON, removes one top-level field
// (used to strip OrgSig / UserSig before signing), and returns the
// JCS-canonical form of the rest.
//
// This is the "signing input" recipe from phase 4:
//
//	sign_bytes(layer, sig_field_name) = utf8(JCS(layer without sig_field_name))
func JCSStripField(v any, fieldName string) ([]byte, error) {
	raw, err := json.Marshal(v)
	if err != nil {
		return nil, fmt.Errorf("json.Marshal: %w", err)
	}
	var m map[string]json.RawMessage
	if err := json.Unmarshal(raw, &m); err != nil {
		return nil, fmt.Errorf("json.Unmarshal to map: %w", err)
	}
	delete(m, fieldName)
	stripped, err := json.Marshal(m)
	if err != nil {
		return nil, fmt.Errorf("json.Marshal stripped: %w", err)
	}
	out, err := jcs.Transform(stripped)
	if err != nil {
		return nil, fmt.Errorf("jcs.Transform: %w", err)
	}
	return out, nil
}
